src/OpenTelemetry.Sampler.AWS/RulesCache.cs

```diff
@@ -60,11 +60,14 @@ public void UpdateRules(List<SamplingRule> newRules)
         List<SamplingRuleApplier> newRuleAppliers = new List<SamplingRuleApplier>();
         foreach (var rule in newRules)
         {
-            var currentStatistics = this.RuleAppliers
-                .FirstOrDefault(currentApplier => currentApplier.RuleName == rule.RuleName)
-                ?.Statistics ?? new Statistics();
+            // If the ruleApplier already exists in the current list of appliers, then we reuse it.
+            var ruleApplier = this.RuleAppliers
+                .FirstOrDefault(currentApplier => currentApplier.RuleName == rule.RuleName) ??
+                new SamplingRuleApplier(this.ClientId, this.Clock, rule, new Statistics());
+
+            // update the rule in the applier in case rule attributes have changed
+            ruleApplier.Rule = rule;
 
-            var ruleApplier = new SamplingRuleApplier(this.ClientId, this.Clock, rule, currentStatistics);
```

Comment:
I recall this was done to "sort of" implement immutability of the `ruleApplier` instances (which is generally a good idea). But as we discussed offline this leads to some race conditions. Can you briefly explain either as a comment in the code or in this PR on why we need to break the immutability and mutate the instance now?

Reply:
This change was made mainly to address one race condition. Basically, with the previous implementation, whenever the RulesCache is periodically updated, it would pull in the list of rules available and if the rule already existed, the Statistics were carried over into a new instance of the rule applier. This caused a race condition. Say the rule had a reservoir of 500 requests/sec and FixedRateSampler with 1/sec after the reservoir is depleted. When creating a new instance of the rule, the constructor creates a new RateLimittingSampler with a reservoir of 1/s and fixedRateSampler with with 1/sec as well since that's in the rule definitio until UpdateTargets is called to set the reservoir according to how the rule is set in cloudwatch (500/s). Between creating a new instance and calling UpdateTargets, if we get a bunch of 200 requests/sec that match the rule, only 2 will be sampled, 1 due to reservoir sampler and the other due to falling back to the fixed rate sampler. If the sampler poling is set to refresh each second, then this will be a problem since it can be the case that the reservoir will mostly only be 1 and not the intended 500. With this change, it will retain the RateLimitingSampler with the correct reservoir while only the rule and it's definition will get updated.

Reply:
Makes sense and lgtm.
Another thing we could have done to preserve immutability was to also pass in the current reservoir size [here](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Sampler.AWS/SamplingRuleApplier.cs#L14) and create a new `RateLimitingSampler` with this reservoir size for a new `SamplingRuleApplier` instance. Just a suggestion, not required right away.